OpenTrons/opentrons-api
# fix(api): enable deck conflict checker for the flex

# Overview

closes https://opentrons.atlassian.net/browse/RSS-294 -> enable `deck_conflict` checker for the flex..
closes https://opentrons.atlassian.net/browse/RSS-273 -> fixes bug when loading a thermocycler in slots that are not allowed. 
closes https://opentrons.atlassian.net/browse/RQA-993 -> does not allow loading mag deck's on an OT-3. just mag blocks. 

# Test Plan

- Allow an OT-3 protocol with modules as described [here](https://opentrons.atlassian.net/browse/RSS-294). make sure no exception is being raise.  
```
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.15",
}

requirements = {
    "robotType": "OT-3"
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    tip_rack_50 = ctx.load_labware("opentrons_ot3_96_tiprack_50ul", 4)
    tip_rack_200 = ctx.load_labware("opentrons_flex_96_tiprack_200ul", 5)
    tip_rack_1000 = ctx.load_labware("opentrons_ot3_96_tiprack_1000ul", 6)
    heatershaker = ctx.load_module('heaterShakerModuleV1',1)
    heatershaker = ctx.load_module('thermocyclerModuleV2')
    mag_block = ctx.load_module('magneticBlockV1',2)
    temp = ctx.load_module('temperatureModuleV2', 3)
```

- Try loading a module on the flex in a middle slot. make sure you are getting an exception. 
```
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.15",
}

requirements = {
    "robotType": "OT-3"
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    # this should fail bc h/s is not allowed in the middle.
    mag_block = ctx.load_module('heaterShakerModuleV1',2)
```
- Try loading 2 modules on the flex in same slot slot. make sure you are getting an exception. 
```
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.15",
}

requirements = {
    "robotType": "OT-3"
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    tip_rack_50 = ctx.load_labware("opentrons_ot3_96_tiprack_50ul", 4)
    tip_rack_200 = ctx.load_labware("opentrons_flex_96_tiprack_200ul", 5)
    tip_rack_1000 = ctx.load_labware("opentrons_ot3_96_tiprack_1000ul", 2)
    heatershaker = ctx.load_module('heaterShakerModuleV1',1)
    # this should fail bc h/s is already loaded in slot 1
    mag_block = ctx.load_module('magneticBlockV1',1)
```

- Should fail bc mag decks are not allowed on the flex.
```
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.15",
}

requirements = {
    "robotType": "OT-3"
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    # this should fail bc h/s is not allowed in the middle.
    mag_deck = ctx.load_module('magneticModuleV2',1)
```

- Should fail bc thermocyclers are not allowed in slot 1.
```
from opentrons import protocol_api, types

metadata = {
    "apiLevel": "2.15",
}

requirements = {
    "robotType": "OT-3"
}

def run(ctx: protocol_api.ProtocolContext) -> None:
    # this should fail bc h/s is not allowed in the middle.
    thermo = ctx.load_module('thermocyclerModuleV2',1)
```

- Smoke test OT-2 protocols and make sure it all works as expected. 
- I have also created an integration test for that
- We should test this with the app and a mag block in a protocol to make sure that it appears properly in the app

# Changelog

- Change `deck_conflict.check` to get DeckSlotName for the new location and and existing location instead of int. 
- Added `robot_type` arg inorder to know what restrictions to load. 
- Changed all access with method to DeckSlotName instead of int.
- Added integration test for a failed protocol
- Updated all tests to use DeckSlotName instead of int/str
- Make sure you are able to load the thermocycler in allowed slots. 
- Added slot validation for load_module in protocol core 

# Review requests

Changes make sense? Did I forget anything? 

# Risk assessment

Medium. need to smoke test OT-2 protocols both json on python for all version (We care about) to make sure loading labware/modules on deck has been affected. need to test OT-3 protocols and to make sure `load_module` is raising an exception for the restrictions in the ticket.
